[Frontend] Add -public-autolink-library option

kateinoigakukun opened the pull request:
Foundation imports CoreFoundation with `@_implementationOnly`,
so CoreFoundation's modulemap won't be read, and the dependent libraries
of CoreFoundation will not be automatically linked when using static
linking.

For example, CoreFoundation depends on libicui18n and it's modulemap has
`link "icui18n"` statement. If Foundation imports CoreFoundation with
`@_implementationOnly` as a private dependency, the toolchain doesn't have
CoreFoundation's modulemap and Foundation's swiftmodule  doesn't import
CoreFoundation. So the swiftc can't know that libicui18n is required.

This new option will add LINK_LIBRARY entry in swiftmodule to
specify dependent libraries (in the example case, Foundation's
swiftmodule should have LINK_LIBRARY entry of libicui18n)


See also: [Autolinking behavior of @_implementationOnly with static linking](https://forums.swift.org/t/autolinking-behavior-of-implementationonly-with-static-linking/44393)

CC: @compnerd

MaxDesiatov commented:
@swift-ci please smoke test

keith commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -1638,6 +1638,10 @@ static bool ParseIRGenArgs(IRGenOptions &Opts, ArgList &Args,
   for (const auto &Lib : Args.getAllArgValues(options::OPT_autolink_library))
     Opts.LinkLibraries.push_back(LinkLibrary(Lib, LibraryKind::Library));
 
+  for (const auto &Lib : Args.getAllArgValues(options::OPT_autolink_library)) {
```
```suggestion
  for (const auto &Lib : Args.getAllArgValues(options::OPT_public_autolink_library)) {
```

? Otherwise this new option doesn't appear to be read

kateinoigakukun replied:
Thanks! My test case was also wrong 😅

MaxDesiatov commented:
@swift-ci please smoke test

buttaface commented:
I've had the same issue on Android for the last couple months.

buttaface commented:
To be clear, the issue for me is that once `CoreFoundation` went `@_implementationOnly` on Android too, the `CoreFoundation` and `uuid` static libraries are no longer linked against from the module map, breaking linking against Foundation statically. Since [the plan is to do the same on linux one day](https://forums.swift.org/t/formalizing-the-unavailability-of-core-foundation/40216), there's going to need to be a solution for static linking against `libCoreFoundation.a` on all non-Darwin platforms eventually.

@drexin or @spevans, could one of you provide direction on this?

kateinoigakukun committed:
[Frontend] Add -public-autolink-library option

Foundation imports CoreFoundation with @_implementationOnly,
so CoreFoundation's modulemap won't be read, and the dependent libraries
of CoreFoundation will not be automatically linked when using static
linking.

For example, CoreFoundation depends on libicui18n and it's modulemap has
`link "icui18n"` statement. If Foundation imports CoreFoundation with
@_implementationOnly as a private dependency, the toolchain doesn't have
CoreFoundation's modulemap and Foundation's swiftmodule  doesn't import
CoreFoundation. So the swiftc can't know that libicui18n is required.

This new option will add LINK_LIBRARY entry in swiftmodule to
specify dependent libraries (in the example case, Foundation's
swiftmodule should have LINK_LIBRARY entry of libicui18n)

kateinoigakukun commented:
@compnerd Could you take a look at this change? This feature is required for https://github.com/apple/swift-corelibs-foundation/pull/2996

MaxDesiatov commented:
@swift-ci please smoke test

spevans commented:
Please test with following PR:
https://github.com/apple/swift-corelibs-foundation/pull/2996
https://github.com/apple/swift-integration-tests/pull/87

@swift-ci please test

kateinoigakukun commented:
Could you re-run the CI again? 🙏

MaxDesiatov commented:
@swift-ci please smoke test

kateinoigakukun commented:
@MaxDesiatov Thanks:) Could you re-trigger with PRs links in `please test` comment to run [cross-repo test](https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#cross-repository-testing)?

- apple/swift-corelibs-foundation#2996
- apple/swift-integration-tests#87

MaxDesiatov commented:
@swift-ci please test
apple/swift-corelibs-foundation#2996
apple/swift-integration-tests#87

MaxDesiatov commented:
@swift-ci please test
apple/swift-corelibs-foundation#2996
apple/swift-integration-tests#87

MaxDesiatov commented:
@swift-ci please test
https://github.com/apple/swift-corelibs-foundation/pull/2996
https://github.com/apple/swift-integration-tests/pull/87

MaxDesiatov commented:
@swift-ci please test
apple/swift-corelibs-foundation#2996

kateinoigakukun commented:
Hmm, ci bot is something wrong ? the GitHub status has not been updated 🤔

MaxDesiatov commented:
Please test with following pull request:
https://github.com/apple/swift-corelibs-foundation/pull/2996

@swift-ci Please test

kateinoigakukun commented:
@compnerd Gentle ping :)

And CI status is not updated, but they have been succeeded.

MaxDesiatov commented:
@swift-ci please test Windows platform

MaxDesiatov commented:
@swift-ci please test Windows platform

kateinoigakukun commented:
Hi, @compnerd :) Could you take a look?

beccadax commented on test/Serialization/autolinking.swift:
```diff
@@ -22,6 +22,11 @@
+
 // Linux uses a different autolinking mechanism, based on
 // swift-autolink-extract. This file tests the Darwin mechanism.
 // UNSUPPORTED: autolink-extract
```
If I'm reading this right, we aren't actually running this test on Linux, the platform where we need this change to work. Is it being properly tested? Is there a different test we could use to verify that it's working on non-Darwin platforms?

kateinoigakukun replied:
Good catch. In theory, this new option is not limited to Linux platform, but should be tested on Linux also. I added a test case to verify it.

MaxDesiatov commented:
@swift-ci please smoke test

CodaFi reviewed (approved):
This is an acceptable solution to an extremely tricky autolinking problem that I imagine we'll encounter in the future in even more exotic configurations than just the core libs.

MaxDesiatov commented:
@swift-ci please smoke test
